h2o-algos/src/main/java/hex/rulefit/Rule.java

```diff
@@ -0,0 +1,170 @@
+        this.languageRule = generateLanguageRule();
+        
+    }
+
+    public void setCoefficient(double coefficient) {
+        this.coefficient = coefficient;
+    }
+
+    String generateLanguageRule() {
+        StringBuilder languageRule = new StringBuilder();
+        if (!this.varName.startsWith("linear.")) {
+            for (int i = 0; i < conditions.length; i++) {
+                if (i != 0)
+                    languageRule.append(" & ");
+                languageRule.append(conditions[i].languageCondition);
+            }
+        }
+        return languageRule.toString();
+    }
+
+    public Frame transform(Frame frame) {
+        Frame frameToReduce = new Frame();
+        for (int i = 0; i < conditions.length; i++) {
+            frameToReduce.add(conditions[i].transform(frame));
+        }
+        RuleReducer mrtask = new RuleReducer();
+        return mrtask.doAll(1, Vec.T_NUM, frameToReduce).outputFrame();
+
+    }
+
+    @Override
+    public int hashCode() {
+        int hashCode = 0;
+        for (int i = 0; i < conditions.length; i++) {
+            hashCode += conditions[i].hashCode();
+        }
+        return hashCode;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        return this.hashCode() == obj.hashCode();
+    }
+
+    public static List<Rule> extractRulesListFromModel(SharedTreeModel model, int modelId) {
+        List<Rule> rules = new ArrayList<>();
+        final SharedTreeModel.SharedTreeOutput sharedTreeOutput = (SharedTreeModel.SharedTreeOutput) model._output;
+        final int treeClass = getResponseLevelIndex(null, sharedTreeOutput);
+        for (int i = 0; i < ((SharedTreeModel.SharedTreeParameters) model._parms)._ntrees; i++) {
+            SharedTreeSubgraph sharedTreeSubgraph = model.getSharedTreeSubgraph(i, treeClass);
+            rules.addAll(extractRulesFromTree(sharedTreeSubgraph, modelId));
+        }
+
+        return rules;
+    }
+    
+    public static Set<Rule> extractRulesFromTree(SharedTreeSubgraph tree, int modelId) {
+        Set<Rule> rules = new HashSet<>();
+        List<Condition> conditions = new ArrayList<>();
+        traverseNodes(tree.rootNode, conditions, rules, null, modelId);
+        return rules;
+    }
+    
+    private static void traverseNodes(SharedTreeNode node, List<Condition> conditions, Set<Rule> rules, Condition conditionToAdd, int modelId) {
+        if (conditionToAdd != null)
+            conditions.add(conditionToAdd);
+        
+        if (node.isLeaf()) {
+            // create Rule to return
+            rules.add(new Rule(conditions.toArray(new Condition[]{}), node.getPredValue(), "M" + modelId + "T" + node.getSubgraphNumber() + "N" + node.getNodeNumber()));
+        } else {
+            // extract this node
+            Condition.Type type;
+            Condition.Operator operator;
+            
+            // extract condition leading to the right and recurse there
+            if (node.getDomainValues() == null) {
```

Comment:
why do you need these two very simmilar if-else statements, I think they can be merged into one, also the `type` and `operator` vars seems unnecessary and can be just removed and values used directly in `new Condition()` calls

Reply:
refactored 63d7852